### PR TITLE
Add string indexing and string hashing functions to stdosl.h

### PR DESCRIPTION
**Pull request by fpsunflower**
These two small functions will be useful from an upcoming contribution to `oslutil.h`.

### CODE REVIEW

**Comment by lgritz** [comment]
LGTM!

I'll add these to the spec after it's merged.

Do you think we should add constant-folded versions of these? Or too much of a rare case to worry about?

**Comment by fpsunflower** [comment]
I think we should add the folders for completeness. Will keep that on my list.

I'm going to at least write a few unit tests.

**Comment by lgritz** [comment]
The low-level string funcs should be tested in testsuite/string/test.osl

Once you've done the magical thing that this is intended for, I would also encourage you to put a new testsuite entry that demonstrates it. (like the one you sent this morning, though low-res is fine)

**Comment by fpsunflower** [comment]
Updated with constant folding and unit tests.

File: src/liboslexec/constfold.cpp [hunk 1]
```diff
@@ -1062,6 +1062,44 @@ DECLFOLDER(constfold_strlen)
 
 
 
+DECLFOLDER(constfold_hash)
+{
+    // Try to turn R=strlen(s) into R=C
```
Comment (lgritz) [inline src/liboslexec/constfold.cpp]: Comment is wrong

File: src/liboslexec/constfold.cpp [hunk 2]
```diff
@@ -1062,6 +1062,44 @@ DECLFOLDER(constfold_strlen)
...
+        return 1;
+    }
+    return 0;
+}
+
+
+
+DECLFOLDER(constfold_getchar)
+{
+    // Try to turn R=strlen(s) into R=C
```
Comment (lgritz) [inline src/liboslexec/constfold.cpp]: Comment is wrong

File: src/liboslexec/constfold.cpp [hunk 3]
```diff
@@ -1062,6 +1062,44 @@ DECLFOLDER(constfold_strlen)
...
 
+DECLFOLDER(constfold_hash)
+{
+    // Try to turn R=strlen(s) into R=C
+    Opcode &op (rop.inst()->ops()[opnum]);
+    Symbol &S (*rop.inst()->argsymbol(op.firstarg()+1));
+    if (S.is_constant()) {
+        ASSERT (S.typespec().is_string());
+        int result = (int) (*(ustring *)S.data()).hash();
+        int cind = rop.add_constant (TypeDesc::TypeInt, &result);
```
Comment (lgritz) [inline src/liboslexec/constfold.cpp]: It's easier to

    int cind = rop.add_constant (result);

It knows result is an int. You only need the (TypeDesc,void*) version when it's not a simple type.

**Comment by lgritz** [comment]
I think you'll also want to update the out.txt output for that test.

**Comment by lgritz** [comment]
Other than my minor comments and the missing reference output, LGTM.

**Comment by fpsunflower** [comment]
Fixed, updated and squashed commits

**Comment by fpsunflower** [comment]
Just because I want to practice my git skills this morning -- update the unit test to properly grab the last character of the string

**Comment by lgritz** [comment]
LGTM

**Pull request merged**